crates/data-structures/src/slim_slice.rs

```diff
@@ -0,0 +1,1934 @@
+
+/// An error signifying that a container's size was over `u32::MAX`.
+///
+/// Optionally, the to-convert object is provided back to the user for handling.
+/// This is what the generic parameter `T` is for.
+#[derive(Error, Debug)]
+#[error("The length `{len}` was too long")]
+pub struct LenTooLong<T = ()> {
+    /// The size of the container that was too large.
+    pub len: usize,
+    /// The container that was too large for into-slim conversion.
+    pub too_long: T,
+}
+
+impl<T> LenTooLong<T> {
+    /// Forgets the container part of the error.
+    pub fn forget(self) -> LenTooLong {
+        self.map(drop)
+    }
+
+    /// Maps the container part of the error.
+    pub fn map<U>(self, with: impl FnOnce(T) -> U) -> LenTooLong<U> {
+        LenTooLong {
+            len: self.len,
+            too_long: with(self.too_long),
+        }
+    }
+}
+
+/// Try to convert `x` into `B` and forget the container part of the error if any.
+#[inline]
+pub fn try_into<A, B: TryFrom<A, Error = LenTooLong<A>>>(x: A) -> Result<B, LenTooLong> {
+    x.try_into().map_err(|e: LenTooLong<A>| e.forget())
+}
+
+// =============================================================================
+// Utils
+// =============================================================================
+
+/// Ensures that `$thing.len() <= u32::MAX`.
+macro_rules! ensure_len_fits {
+    ($thing:expr) => {
+        let Ok(_) = u32::try_from($thing.len()) else {
+            return Err(LenTooLong {
+                len: $thing.len(),
+                too_long: $thing,
+            });
+        };
+    };
+}
+
+/// Convert to `A` but panic if `x.len() > u32::MAX`.
+#[inline]
+fn expect_fit<A, E, B: TryInto<A, Error = LenTooLong<E>>>(x: B) -> A {
+    x.try_into().map_err(|e| e.len).expect("length didn't fit in `u32`")
+}
+
+#[inline]
+fn into_box<T, U: Into<Box<[T]>>>(x: U) -> Box<[T]> {
+    x.into()
+}
+
+// Asserts, in the type system, that `N <= u32::MAX`.
+struct AssertU32<const N: usize>;
+impl<const N: usize> AssertU32<N> {
+    const OK: () = assert!(N <= u32::MAX as usize);
+}
+
+// =============================================================================
+// Raw slice utility type
+// =============================================================================
+
+/// Implementors decree that `Self` can be *safely* transmuted to `T`,
+/// including covariantly under a pointer.
+///
+/// # Safety
+///
+/// It is not sufficient that ´Self` and `T` have the same representation.
+/// That is, validity requirements are not enough.
+/// The safety requirements must also be the same.
+pub unsafe trait SafelyExchangeable<T> {}
+
+#[inline]
+fn transmute_safely_exchangeable<T, U: SafelyExchangeable<T>>(x: T) -> U {
+    let x = ManuallyDrop::new(x);
+    // SAFETY: Caller has proven that `x` is valid for `U`.
+    // The reference is also properly aligned
+    // and valid for reads by virtue of `&*x`.
+    unsafe { read(&*x as *const T as *const U) }
```

Comment:
Any reason not to use transmute_copy? Seems more in-line with the operation being done here.

Reply:
Why is this not just `mem::transmute(x)`?